subscription_manager.go

```diff
@@ -0,0 +1,271 @@
+		event = &Event{Kind: EventKindJobCompleted, Job: job, JobStats: stats}
+	case rivertype.JobStateScheduled:
+		event = &Event{Kind: EventKindJobSnoozed, Job: job, JobStats: stats}
+	case rivertype.JobStateAvailable, rivertype.JobStateDiscarded, rivertype.JobStateRetryable, rivertype.JobStateRunning:
+		event = &Event{Kind: EventKindJobFailed, Job: job, JobStats: stats}
+	case rivertype.JobStatePending:
+		panic("completion subscriber unexpectedly received job in pending state, river bug")
+	default:
+		// linter exhaustive rule prevents this from being reached
+		panic("unreachable state to distribute, river bug")
+	}
+
+	// All subscription channels are non-blocking so this is always fast and
+	// there's no risk of falling behind what producers are sending.
+	for _, sub := range sm.subscriptions {
+		if sub.ListensFor(event.Kind) {
+			// TODO: THIS IS UNSAFE AND WILL LEAD TO DROPPED EVENTS.
+			//
+			// We are allocating subscriber channels with a fixed size of 1000, but
+			// potentially processing job events in batches of 5000 (batch completer
+			// max batch size). It's probably not possible for the subscriber to keep
+			// up with these bursts.
+			select {
+			case sub.Chan <- event:
+			default:
+			}
+		}
+	}
+}
+
+func (sm *subscriptionManager) distributeQueueEvent(event *Event) {
+	sm.mu.Lock()
+	defer sm.mu.Unlock()
+
+	// All subscription channels are non-blocking so this is always fast and
+	// there's no risk of falling behind what producers are sending.
+	for _, sub := range sm.subscriptions {
+		if sub.ListensFor(event.Kind) {
+			select {
+			case sub.Chan <- event:
+			default:
+			}
+		}
+	}
+}
+
+// Special internal variant that lets us inject an overridden size.
+func (sm *subscriptionManager) SubscribeConfig(config *SubscribeConfig) (<-chan *Event, func()) {
+	if config.ChanSize < 0 {
+		panic("SubscribeConfig.ChanSize must be greater or equal to 1")
+	}
+	if config.ChanSize == 0 {
+		config.ChanSize = subscribeChanSizeDefault
+	}
+
+	for _, kind := range config.Kinds {
+		if _, ok := allKinds[kind]; !ok {
+			panic(fmt.Errorf("unknown event kind: %s", kind))
+		}
+	}
+
+	subChan := make(chan *Event, config.ChanSize)
+
+	sm.mu.Lock()
+	defer sm.mu.Unlock()
+
+	// Just gives us an easy way of removing the subscription again later.
+	subID := sm.subscriptionsSeq
+	sm.subscriptionsSeq++
+
+	sm.subscriptions[subID] = &eventSubscription{
+		Chan:  subChan,
+		Kinds: sliceutil.KeyBy(config.Kinds, func(k EventKind) (EventKind, struct{}) { return k, struct{}{} }),
+	}
+
+	cancel := func() {
+		sm.mu.Lock()
+		defer sm.mu.Unlock()
+
+		// May no longer be present in case this was called after a stop.
+		sub, ok := sm.subscriptions[subID]
+		if !ok {
+			return
+		}
+
+		close(sub.Chan)
+
+		delete(sm.subscriptions, subID)
+	}
+
+	return subChan, cancel
+}
+
+// func newAsyncCompleterWithConcurrency(archetype *baseservice.Archetype, exec PartialExecutor, concurrency int, subscribeCh SubscribeChan) *AsyncCompleter {
```

Comment:
Lots of dead code here. Maybe copy/paste error?

Reply:
Yep shouldn't have made it into the commit, sorry about that!